### DISCUSSION TITLE
My Tutorial 1 QEMU output is different (wrong?)

### DISCUSSION THREAD

**Discussion by grtwje** [rust-embedded/rust-raspberrypi-OS-tutorials#92] [Tutorial Questions] [answered]
First, let me start by thanking you for providing this tutorial.

I have a slightly odd set-up. It is Ubuntu under Window Services for Linux 2. I had to install a few packages, but it seems to work.  All of the make targets build without error. But the output I get from 'make qemu' does not match what is given in the README file.  Maybe my set-up is not quite correct?  Here's the output I see:

```
 $ make qemu
 RUSTFLAGS="-C link-arg=-Tsrc/bsp/raspberrypi/link.ld -C target-cpu=cortex-a53 -D warnings -D missing_docs" cargo rustc --target=aarch64-unknown-none-softfloat --features bsp_rpi3 --release
    Compiling kernel v0.1.0 (/mnt/f/Projects/rust_raspberrypi_OS_tutorial/01_wait_forever)
     Finished release [optimized] target(s) in 3.79s
 Unable to find image 'rustembedded/osdev-utils:latest' locally
 latest: Pulling from rustembedded/osdev-utils
 6a5697faee43: Pull complete 
 ba13d3bc422b: Pull complete 
 a254829d9e55: Pull complete 
 ef2df80894b7: Pull complete 
 eabb35eed9af: Pull complete 
 aceb1fd9098d: Pull complete 
 9d3b9a6cac07: Pull complete 
 3d68be893534: Pull complete 
 e299ec6b0db2: Pull complete 
 Digest: sha256:2f4f14db5cba2090ce9120d3d7e1e1fcf9fbac9cfd9e98fc4d7559901eda0163
 Status: Downloaded newer image for rustembedded/osdev-utils:latest
 ----------------
 IN: 
 0x00000300:  d2801b05  movz     x5, #0xd8
 0x00000304:  d53800a6  mrs      x6, mpidr_el1
 0x00000308:  924004c6  and      x6, x6, #3
 0x0000030c:  d503205f  wfe      
 0x00000310:  f86678a4  ldr      x4, [x5, x6, lsl #3]
 0x00000314:  b4ffffc4  cbz      x4, #0x30c
 
 ----------------
 IN: 
 0x0000030c:  d503205f  wfe      
 0x00000310:  f86678a4  ldr      x4, [x5, x6, lsl #3]
 0x00000314:  b4ffffc4  cbz      x4, #0x30c
 
 ----------------
 IN: 
 0x00000000:  580000c0  ldr      x0, #0x18
 0x00000004:  aa1f03e1  mov      x1, xzr
 0x00000008:  aa1f03e2  mov      x2, xzr
 0x0000000c:  aa1f03e3  mov      x3, xzr
 0x00000010:  58000084  ldr      x4, #0x20
 0x00000014:  d61f0080  br       x4
 
 ----------------
 IN: 
 0x0000030c:  d503205f  wfe      
 0x00000310:  f86678a4  ldr      x4, [x5, x6, lsl #3]
 0x00000314:  b4ffffc4  cbz      x4, #0x30c
 
 ----------------
 IN: 
 0x00080000:  00000004  .byte    0x04, 0x00, 0x00, 0x00
 
 ----------------
 IN: 
 0x00000200:  00000000  .byte    0x00, 0x00, 0x00, 0x00
```

**Comment by andre-richter** [comment]
It’s been some time since I wrote this Readme, it might not be the exact same since the time I took a snapshot, because code might have changed slightly here or there.

Does “make qemu” produce meaningful output in tutorials that write to the console, like 04 or 05?
If so, I would say you don’t need to worry that 01 or 02 don’t reproduce exactly.

**Reply by grtwje** [reply]
Here's the output for `make qemu` for those 2 tutorials.  I have not studied those tutorials to know what is expected.  But the output looks like it may be correct.

Tutorial 4
```
[0] Hello from pure Rust!
[1] Stopping here.
```
Tutorial 5
```
[0] Hello from pure Rust!
[1] Chars written: 27
[2] Stopping here.
```

**Comment by andre-richter** [comment, marked as answer]
Upon further inspection, it turns out that you found a bug that was introduced when I added LTO to tutorial 01.
For some reason, it messed with the address of the `_start` function, causing it to not be linked at `0x80000`.

That is why your QEMU showed that no actual `AArch64` instruction was living at `0x80000`:
```
0x00080000:  00000004  .byte    0x04, 0x00, 0x00, 0x00
```

Commit https://github.com/rust-embedded/rust-raspberrypi-OS-tutorials/commit/53c1163c51931ac5c58da03cffeb46ada2486f22 fixed this.
Now you should be able to see the `wfe` infinite loop starting at `0x80000` like shown in the Readme:

```
IN:
0x00080000:  d503205f  wfe
0x00080004:  17ffffff  b        #0x80000
```

Thanks for reporting this!
Andre
